installer/install.py

```diff
@@ -268,42 +266,70 @@ def debug_wifi():
 
 report_success()
 
-debug_wifi()
+os.system("/userdata/x1plus/debug_wifi.sh")
+
+def download_firmware(update_url, dest_path):
+    headers = {
+        'User-Agent': f"Python-urllib/3.{sys.version_info.minor}"
+    }
+    
+    os.makedirs(os.path.dirname(dest_path), exist_ok=True)
+
+    report_interim_progress("Connecting...")
+    success = False
+    retry = 1
+    while success == False and retry < 20:
+        try:
+            resp = requests.get(update_url, headers=headers, stream=True, verify="/etc/ssl/certs/")
+            resp.raise_for_status()
+        except (requests.Timeout, requests.exceptions.HTTPError, requests.exceptions.RequestException) as e:
+            report_interim_progress(f"{e.__class__.__name__}: {e} ... retry {retry}")
+            time.sleep( pow( 2, retry ) if retry < 4 else 16 )
+            retry += 1
+        except Exception as e:
+            report_interim_progress(f"An error occurred: {e} ... retry {retry}")
+            time.sleep( pow( 2, retry ) if retry < 4 else 16 )
+            retry += 1
+        else:
+            success = True
```

Comment:
This retries if the initial connection setup fails, but not if the download later fails (i.e., if we download fewer than `totlen` bytes), right?  Is this the desired behavior?

Reply:
This could be avoided for the most part by running `ping` before the download is attempted. If the ping fails, then don't download. This is also a good idea in general, because we don't really know what our margins for error are on timing for avoiding this race condition. It's not exactly a race condition but you know what I mean.

Reply:
there are other errors that can occur even if ping works. we can add totlen checking as well.

Reply:
Right, the only thing this can serve as a breakpoint for is when a printer is able to ping google but not connect to Bambu's CDN links. 

After the issues we've been having, I would say the more breakpoints the better